### ISSUE TITLE
Zero diskIO write stats on systems using dm-crypt and the CFQ scheduler

### ISSUE THREAD

**Issue by saj** [google/cadvisor#2316]
Given an I/O device stack that looks something like this:

```
$ lsblk
NAME                    MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda                       8:0    0 953.9G  0 disk
├─sda1                    8:1    0     4M  0 part
├─sda2                    8:2    0   512M  0 part  /boot
└─sda3                    8:3    0 953.4G  0 part
  └─pv_crypt            254:0    0 953.4G  0 crypt
    ├─redacted-root 254:1    0 890.6G  0 lvm   /
    └─redacted-swap 254:2    0  62.8G  0 lvm   [SWAP]
```

Any **writes** made by containerised workloads to dm-crypt encrypted block devices are not observed or exported by cadvisor.  (Device `254:1` in the snippet above would be an example of such a block device.)

All **reads** appear to be correctly observed and exported.

I think this bug is the result of the following:

- cadvisor will always prefer to source data from the `blkio.io_service*` pseudofiles in the cgroup VFS when the CFQ scheduler observes at least one I/O.  (See commit: [Use blkio.throttle.* stats when CFQ is not in use](https://github.com/opencontainers/runc/commit/78700bfea36558fc3d32ca909bff9317fe791e4d).)

  These pseudofiles only report on I/O to _physical_ disk devices as it is on those devices where the CFQ scheduler is attached.  I/O to dm devices does not appear here.

- The Linux kernel (apparently) drops group context from all writes that traverse a dm-crypt device.

Typical sequence of events:

(where _container_ is any process executing in a container cgroup)

1. Container starts.

1. Container reads data from a filesystem on a dm-crypt device, triggering a miss on the kernel page cache.  The data is paged in from disk.  Because the kernel does correctly maintain group context on reads, the `blkio.io_service*` pseudofiles begin reporting non-zero counters for reads from physical devices only.  cadvisor begins exporting from `blkio.io_service*`.

1. Container writes data to a filesystem on a dm-crypt device.  It does not matter how these writes are made.  (Async or sync, the end result is the same.)  Because the kernel does _not_ maintain group context on writes that traverse a dm-crypt device, the `blkio.io_service*` pseudofiles continue to report zero write operations.  Meanwhile, the write counters in the `blkio.throttle.io_service*` pseudofiles begin incrementing as expected for dm devices.  This is moot, though, because cadvisor ignores this data in preference for the (significantly more incomplete) stats from the CFQ scheduler.

I would suppose this is technically a bug in the kernel, though a userspace workaround may be the only practical path forward in the near term.

dm-crypt users must make peace with the fact that we will not be able to attribute _physical_ block device writes to discrete containers.  The data is simply not available -- neither in `blkio.io_service*` nor in `blkio.throttle.io_service*`.  That aside, I think the current implementation of cadvisor unnecessarily denies us access to useful _virtual_ block device I/O metrics, like writes to dm-crypt and LVM devices.

May I ask what problem the [aforementioned commit](https://github.com/opencontainers/runc/commit/78700bfea36558fc3d32ca909bff9317fe791e4d) had intended to solve?  (There is a bug ref in the commit message, though that bug report seems unrelated to the patch.)  **What is the benefit of preferencing data from the CFQ scheduler?**

We have observed this problem on Linux 4.9 I/O device stacks that incorporate a dm-crypt layer, though it is possible that the general problem of lost group context impacts users of _other_ device-mapper magic as well.  Bug reports #1169, #1905, #1946, and #2148 may be related.

**Comment by saj** [comment]
Oh, I forgot to make this explicit -

There is _one_ case where we do successfully get write I/O metrics out of cadvisor.  Some of our workloads do not read data from disk on startup.  (Or whatever data they need is already in the page cache.)  From that point onwards, these workloads only write to the filesystem.  Because there are no reads, the `blkio.io_service*` pseudofiles (from CFQ) never report any I/Os, and cadvisor is able to fall-back to reading data from the `blkio.throttle.io_service*` pseudofiles.  In these cases we are able to observe read and write I/Os to our LVM and dm-crypt block devices.

Given kernel constraints, this is ideal behaviour, and is what I'm after for all workloads.

**Comment by dashpole** [comment]
I'm afraid I don't have context on why that decision was made initially, as that was before my time.

**Comment by szediktam** [comment]
https://github.com/google/cadvisor/issues/3047
@saj  @dashpole  do you have any idea about this issue? thanks.

**Comment by saj** [comment]
@szediktam, not really.  This bug (#2316) impacted _write_ metrics, not _read_ metrics.  IIRC, we never experienced a problem with read metrics.

> I create two pods with same config, container in a pod has metric container_fs_reads_bytes_total, but the other has no this metric. Why is that?

Are you certain your workload is actually reading from disk?  A hit on the page cache will obviate the need for disk I/O.

**Comment by saj** [comment]
I have an update on the original problem.  At the time I filed this bug, most of our production systems were running Debian 9 (Linux 4.9) with the CFQ scheduler.  Later, those systems were updated to Debian 10 (Linux 4.19).  We continue to use dm-crypt.  Somewhere along the line -- perhaps it is a distro default -- these systems transitioned from the CFQ scheduler to the mq-deadline scheduler.  This problem went away.

I was forced to fork runc and cadvisor while on Debian 9.  Our builds disabled the `getCFQStats` function.  I dropped that fork, and returned to upstream builds, after Debian 10.

From my notes:

> `blkio.io_serviced`: reports the number of I/O operations performed on specific devices by a cgroup **as seen by the CFQ scheduler**.
>
> `blkio.throttle.io_serviced`: reports the number of I/O operations performed on specific devices by a cgroup **as seen by the throttling policy**.

Previously, on Debian 9:

* reads were instrumented in both `blkio.io_serviced` (for phy devices) and `blkio.throttle.io_serviced` (for virt devices)
* writes were only instrumented in `blkio.throttle.io_serviced`
* this R/W asymmetry was problematic with cadvisor/runc because if any data was found in `blkio.io_serviced`, then cadvisor would preference that data source for read and write metrics – even though no writes would ever appear there

On Debian 10, no data ever appears in `blkio.io_serviced`:

*  **reads do not appear in `blkio.io_serviced`**
*  asynchronous writes do not appear in `blkio.io_serviced`
   (`dd if=/dev/zero of=/saj/1-mib count=64 bs=16K`)
*  writes followed by an `fsync()` do not appear in `blkio.io_serviced`
   (`dd if=/dev/zero of=/saj/1-mib count=64 bs=16K conv=fsync`)
*  writes that bypass the page cache do not appear in `blkio.io_serviced`
   (`dd if=/dev/zero of=/saj/1-mib count=64 bs=16K oflag=direct`)

This is quite fortuitous because I think it means we can drop our fork and return to upstream cadvisor builds. The convoluted maze of factors that caused the original problem has shifted just enough to clear a path. The difference between our old Debian 9 and new Debian 10 systems may lie in our chosen I/O scheduler.

As before, series for physical devices contain incomplete, and potentially misleading, data and should not be used. (e.g.: /dev/nvme0n1 in the screenshot above.) If you need data at that layer, you must use the node_ metrics (which do not aggregate by cgroup - too bad).

**Comment by szediktam** [comment]
> @szediktam, not really. This bug (#2316) impacted _write_ metrics, not _read_ metrics. IIRC, we never experienced a problem with read metrics.
> 
> > I create two pods with same config, container in a pod has metric container_fs_reads_bytes_total, but the other has no this metric. Why is that?
> 
> Are you certain your workload is actually reading from disk? A hit on the page cache will obviate the need for disk I/O.
the workload may be not read fron disk.
Bug i think the two container should have the same behavior: both have the metric, or both have no metric. Now one has metric with value 0, the other has no metric. I don't understand this.

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
